### PR TITLE
Refactored @puppy3276's plugin pull request

### PR DESCRIPTION
**Pull request by pdelvo**
Refactored version of pull request #6 by @puppy3276

### CODE REVIEW

**Comment by ddevault** [comment]
ack, ack, ack, conflicting pull requests

**Comment by pdelvo** [comment]
@puppy3276 closed his pull request seconds ago

**Comment by ddevault** [comment]
Yep, looking through this one now.

File: PartyCraft/PluginCore.cs [hunk 1]
```diff
@@ -0,0 +1,348 @@
...
+        public void LoadPlugins(string folder)
+        {
+            Log("Begin loading plugins");
+            BeginSubLog();
+
+            // First check if the core isn't already disposed
+            if (this.IsDisposed)
+            {
+                throw new ObjectDisposedException("The object has already been disposed");
+            }
```
Comment (ddevault) [inline PartyCraft/PluginCore.cs]: Single-line if statements shouldn't be enclosed in brackets

File: PartyCraft/PluginCore.cs [hunk 2]
```diff
@@ -0,0 +1,348 @@
...
+                // This is where we will store all the IPlugins available in the DLL
+                List<IPluginSystem> plugins = new List<IPluginSystem>();                                        
+                
+                // Later we will load the plugin using pluginName which is without the DLL
+                string pluginName = System.IO.Path.GetFileNameWithoutExtension(pluginFilename);
+                
+                // Creating the domain. Important is that we define the folder using AppDomainSetup. 
+                // It's important that we load the plugins in another application domain as we should see it as
+                // an extension and not part of the current application.
+                System.Security.Policy.Evidence evidence = new System.Security.Policy.Evidence();
```
Comment (ddevault) [inline PartyCraft/PluginCore.cs]: Add a using statement instead

File: PartyCraft/PluginCore.cs [hunk 3]
```diff
@@ -0,0 +1,348 @@
...
+                appDomainSetup.ApplicationBase = folder;
+                AppDomain pluginDomain = AppDomain.CreateDomain("Plugin: " + pluginFilename, evidence, appDomainSetup);
+
+                // This list will contain all the class that have an IPlugin interface
+                List<Type> pluginTypes = null;
+
+                try
+                {
+                    // Loading the assembly
+                    System.Reflection.Assembly pluginAssembly = pluginDomain.Load(pluginName);
```
Comment (ddevault) [inline PartyCraft/PluginCore.cs]: Again, add a using statement

File: PartyCraft/PluginCore.cs [hunk 4]
```diff
@@ -0,0 +1,348 @@
...
+
+                        // If the constructor doesn't exist, we have an IPlugin object who doesn't match the requirement. 
+                        // Let's find out the reason why.
+                        if (pluginConstructor == null)
+                        {
+                            // Check if the constructor is private (one might have forgotten the public keyword). It's 
+                            // a separate check as this is a common problem. The first time I forgot I spend a good 5 
+                            // minutes before I figured it out.
+                            var privatePluginConstructor = pluginType.GetConstructors(System.Reflection.BindingFlags.NonPublic | System.Reflection.BindingFlags.Instance)
+                                                                     .Where(_con => _con.GetParameters().Count() == 1).SingleOrDefault(_con => _con.GetParameters()[0].ParameterType == typeof(PluginCore));
```
Comment (ddevault) [inline PartyCraft/PluginCore.cs]: Continuing a line on another line should be indented once.

```
var privatePluginConstructor = pluginType.GetConstructors(System.Reflection.BindingFlags.NonPublic | System.Reflection.BindingFlags.Instance)
    .Where(_con => _con.GetParameters().Count() == 1).SingleOrDefault(_con => _con.GetParameters()[0].ParameterType == typeof(PluginCore));
```

Again, add using statements instead of fully declaring type names.

File: PartyCraft/PluginCore.cs [hunk 5]
```diff
@@ -0,0 +1,348 @@
...
+                    // that we find the assembly that manages the plugins. We should ignore that one.
+                    if (pluginAssembly.GetTypes().Contains(typeof(PluginCore)))
+                    {
+                        Log("'{0}' is considered the SDK as PluginCore is detected, skipping the assembly", pluginName);
+                        AppDomain.Unload(pluginDomain); // Unload domain as we are not using it
+                        continue;
+                    }
+
+                    // Find all plugin
+                    pluginTypes = pluginAssembly.GetTypes().Where(t => t.GetInterfaces().Contains(typeof(IPluginSystem))).ToList();
```
Comment (ddevault) [inline PartyCraft/PluginCore.cs]: Abstract IPlugin classes aren't handled here.

File: PartyCraft/PluginCore.cs [hunk 6]
```diff
@@ -0,0 +1,348 @@
...
+                                // There is no private constructor (and we ruled out a public constructor). Only option is that is not correctly defined
+                                throw new Exception(String.Format("The plugin '{0}' defined in '{1}.dll' has no constructor that accepts PluginCore as only parameter and can thus not be initialized. Please define the constructor that matches those requirements", pluginType.FullName, pluginName));
+                            }
+                        }
+
+                        // Now lets invoke the constructor and instantiate an IPlugin
+                        IPluginSystem plugin = pluginConstructor.Invoke(new Object[] { this }) as IPluginSystem;
+
+                        // Now check if the plugin was already loaded (let's prevent duplicate plugins, plugins are already trouble enough)
+                        if (hosts.Any(loadedPlugin => loadedPlugin.Plugin.GetName() == plugin.GetName()) || plugins.Any(loaded_plugin => loaded_plugin.GetName() == plugin.GetName()))
```
Comment (ddevault) [inline PartyCraft/PluginCore.cs]: Use GUIDs instead.

File: PartyCraft/PluginCore.cs [hunk 7]
```diff
@@ -0,0 +1,348 @@
...
+                        plugins.Add(plugin);
+                        Log("{0} from {1} was loaded as plugin", plugin.GetName(), pluginName);
+                    }
+
+                    // At this point all plugins from the assembly have been loaded (if one plugin failed, the entire assembly failed).
+                    // Now let's store it in a host.
+                    foreach (var plugin in plugins)
+                    {
+                        hosts.Add(new PluginHost(this, plugin, pluginDomain));
+                    }
```
Comment (ddevault) [inline PartyCraft/PluginCore.cs]: No braces here, please

File: PartyCraft/PluginCore.cs [hunk 8]
```diff
@@ -0,0 +1,348 @@
...
+                }
+            }
+            finally
+            {
+                // explicitly call the base class Dispose implementation
+                this.IsDisposed = true;
+            }
+        }
+        
+        //TODO Uncomment this code if this class will contain members which are UNmanaged
```
Comment (ddevault) [inline PartyCraft/PluginCore.cs:274]: This class likely won't ever have unmanaged members, why is this here?

File: PartyCraft/Program.cs [hunk 9]
```diff
@@ -21,12 +21,25 @@ public class Program
         /// </summary>
         public static ISettingsProvider SettingsProvider = null;
 
+        private static string pluginDirectory = "plugins";
```
Comment (ddevault) [inline PartyCraft/Program.cs]: Organize it better - keep public/private, static/instance fields and properties separate. Also, make this a constant.

File: PartyCraft/TextLogger.cs [hunk 10]
```diff
@@ -0,0 +1,92 @@
...
+using System.Collections.Generic;
+using System.Linq;
+using System.Text;
+
+namespace PartyCraft
+{
+    public class TextLogger : ILogger, IDisposable
+    {
+        int level = 0;
+        private String m_Filename;
```
Comment (ddevault) [inline PartyCraft/TextLogger.cs]: 'string', not 'String', and m_ is unacceptable. Try this:

```
public string Filename { get; private set; }
```

File: PartyCraft/PluginCore.cs (hunk 6 above)
Comment (pdelvo) [inline PartyCraft/PluginCore.cs]: Or comparing the type ;)

File: PartyCraft/TextLogger.cs [hunk 11]
```diff
@@ -0,0 +1,92 @@
...
+        public void BeginSubLog()
+        {
+            if (isDisposed) throw new ObjectDisposedException("The object has already been disposed");
+            level++;
+        }
+
+        public void EndSubLog()
+        {
+            if (isDisposed) throw new ObjectDisposedException("The object has already been disposed");
+            level--;
```
Comment (ddevault) [inline PartyCraft/TextLogger.cs:37]: Can't this be done better with a `System.Collections.Generics.Stack<T>`?

**Comment by ghost** [comment]
Nevermind.

File: PartyCraft/TextLogger.cs [hunk 12]
```diff
@@ -0,0 +1,92 @@
...
+        public int GetSubLogDepth()
+        {
+            if (isDisposed) throw new ObjectDisposedException("The object has already been disposed");
+            return level;
+        }
+
+        public void Log(string message)
+        {
+            if (isDisposed) throw new ObjectDisposedException("The object has already been disposed");
+            String log_text = "[" + DateTime.Now.ToString("yyyyMMdd HH:mm:ss") + "]";
```
Comment (ddevault) [inline PartyCraft/TextLogger.cs]: This produces things like `20120712` for today's date, add better formatting.

File: PluginSystemLib/PluginCore.cs [hunk 13]
```diff
@@ -0,0 +1,360 @@
+﻿using System;
```
Comment (ddevault) [inline PluginSystemLib/PluginCore.cs]: Isn't this file a duplicate?

File: PluginSystemLib/PluginCoreConfiguration.cs [hunk 14]
```diff
@@ -0,0 +1,7 @@
+﻿namespace PluginSystemCS
```
Comment (ddevault) [inline PluginSystemLib/PluginCoreConfiguration.cs]: More duplicates follow.

File: PluginSystemLib/TextLogger.cs [hunk 15]
```diff
@@ -0,0 +1,92 @@
+﻿using System;
```
Comment (ddevault) [inline PluginSystemLib/TextLogger.cs]: Oh, there's this PluginSystemLib project, why is this separate from the rest?

File: PartyCraft/TextLogger.cs (hunk 11 above)
Comment (pdelvo) [inline PartyCraft/TextLogger.cs:37]: It can be done with it, but I dont think this would be better. KISS - Keep it short and simple

File: lib/docs/LICENSE [hunk 16]
```diff
@@ -0,0 +1,165 @@
+		   GNU LESSER GENERAL PUBLIC LICENSE
```
Comment (ddevault) [inline lib/docs/LICENSE:1]: AHAHAHAHAHA NO

File: PartyCraft/TextLogger.cs (hunk 12 above)
Comment (pdelvo) [inline PartyCraft/TextLogger.cs]: I remove the format part. Logging in the current culture format should be the result the user expect

**Comment by ddevault** [comment]
Comments:
- Don't just generate docs, better to not include them so there are warnings to go back and add it later
- LibNbt is in here, get it out of here. It's a submodule of Craft.Net
- What's this PluginSystemLib thing? Firstly, that's a bad project name, secondly, it duplicates a ton of stuff from the PartyCraft changes.

File: PluginSystemLib/TextLogger.cs (hunk 15 above)
Comment (pdelvo) [inline PluginSystemLib/TextLogger.cs]: I did not even know it exists. It was also in the partycraft project.

File: PartyCraft/PluginCore.cs (hunk 6 above)
Comment (ddevault) [inline PartyCraft/PluginCore.cs]: Or that, sure, but it'd be nice if all plugins had a GUID anyway.

File: PartyCraft/TextLogger.cs (hunk 11 above)
Comment (ddevault) [inline PartyCraft/TextLogger.cs:37]: I like the idea of having a stack of logs. I'll just refactor this myself after merging this request.

File: PluginSystemLib/TextLogger.cs (hunk 15 above)
Comment (ddevault) [inline PluginSystemLib/TextLogger.cs]: Weird.

**Comment by ddevault** [comment]
Also, it's probably best to revert all the changes on `lib/` in the repo.

**Comment by pdelvo** [comment]
I stop fixing it. @puppy3276 you can go on now!
